ibis/backends/datafusion/compiler.py

```diff
@@ -0,0 +1,410 @@
+import functools
+import operator
+
+import datafusion as df
+import datafusion.functions
+import pyarrow as pa
+
+import ibis.common.exceptions as com
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+
+from .datatypes import to_pyarrow_type
+
+
+@functools.singledispatch
+def translate(expr):
+    raise NotImplementedError(expr)
+
+
+@translate.register(ir.Expr)
+def expression(expr):
+    return translate(expr.op(), expr)
+
+
+@translate.register(ops.Node)
+def operation(op, expr):
+    raise com.OperationNotDefinedError(f'No translation rule for {type(op)}')
+
+
+@translate.register(ops.DatabaseTable)
+def table(op, expr):
+    name, _, client = op.args
+    return client.context.table(name)
+
+
+@translate.register(ops.Literal)
+def literal(op, expr):
+    if isinstance(op.value, (set, frozenset)):
+        value = list(op.value)
+    else:
+        value = op.value
+
+    arrow_type = to_pyarrow_type(op.dtype)
+    arrow_scalar = pa.scalar(value, type=arrow_type)
+
+    return df.literal(arrow_scalar)
+
+
+@translate.register(ops.Cast)
+def cast(op, expr):
+    arg = translate(op.arg)
+    typ = to_pyarrow_type(op.to)
+    return arg.cast(to=typ)
+
+
+@translate.register(ops.TableColumn)
+def column(op, expr):
+    table_op = op.table.op()
+
+    if hasattr(table_op, "name"):
+        return df.column(f'{table_op.name}."{op.name}"')
+    else:
+        return df.column(op.name)
+
+
+@translate.register(ops.SortKey)
+def sort_key(op, expr):
+    arg = translate(op.expr)
+    return arg.sort(ascending=op.ascending)
+
+
+@translate.register(ops.Selection)
+def selection(op, expr):
+    plan = translate(op.table)
+
+    selections = []
+    for expr in op.selections or [op.table]:
+        # TODO(kszucs) it would be nice if we wouldn't need to handle the
+        # specific cases in the backend implementations, we could add a
+        # new operator which retrieves all of the TableExpr columns
+        # (.e.g. Asterisk) so the translate() would handle this
+        # automatically
+        if isinstance(expr, ir.TableExpr):
+            for name in expr.columns:
+                column = expr.get_column(name)
+                field = translate(column)
+                if column.has_name():
+                    field = field.alias(column.get_name())
+                selections.append(field)
+        elif isinstance(expr, ir.ValueExpr):
+            field = translate(expr)
+            if expr.has_name():
+                field = field.alias(expr.get_name())
+            selections.append(field)
+        else:
+            raise com.IbisError("TODO")
+
+    plan = plan.select(*selections)
+
+    if op.predicates:
+        predicates = map(translate, op.predicates)
+        predicate = functools.reduce(operator.add, predicates)
```

Comment:
This should be `and`, not `add` :sweat_smile:

Reply:
Probably needs a test too, if this isn't failing.

Reply:
bump

Reply:
Nice catch! Updated and added a test case.